feat: add retention config for `workspace_agent_logs`

mafredri opened the pull request:
Replace hardcoded 7-day retention for workspace agent logs with configurable
retention from deployment settings. Defaults to 7d to preserve existing behavior.

Depends on #21038
Updates #20743

---

### PR Stack
| | PR | Title |
|:-:|:-:|:--|
| | #21021 | feat(coderd): add retention policy configuration |
| | #21022 | feat(coderd/database/dbpurge): add retention for connection logs |
| | #21025 | feat(coderd/database/dbpurge): add retention for audit logs |
| | #21037 | feat(coderd/database/dbpurge): make API keys retention configurable |
| | #21038 | docs: add data retention documentation |
| 👉 | #21039 | feat: add retention config for `workspace_agent_logs` |

mafredri committed:
feat(coderd/database/dbpurge): add retention for workspace agent logs

Replace hardcoded 7-day retention for workspace agent logs with configurable
retention from deployment settings. Falls back to global retention when not
set, and skips deletion entirely when effective retention is 0.

Depends on #21038
Updates #20743

mafredri committed:
chore: remove global retention fallback for workspace agent logs

Workspace agent logs retention is now explicit - it's enabled when
--workspace-agent-logs-retention is set to a non-zero duration
(default 7d), and disabled when set to 0. No fallback to global
retention.

dannykopping commented on coderd/database/dbpurge/dbpurge.go:
```diff
@@ -67,9 +66,12 @@ func New(ctx context.Context, logger slog.Logger, db database.Store, vals *coder
+			workspaceAgentLogsRetention := vals.Retention.WorkspaceAgentLogs.Value()
+			if workspaceAgentLogsRetention > 0 {
+				deleteOldWorkspaceAgentLogsBefore := start.Add(-workspaceAgentLogsRetention)
+				if err := tx.DeleteOldWorkspaceAgentLogs(ctx, deleteOldWorkspaceAgentLogsBefore); err != nil {
```
Please log the count to remain consistent.

dannykopping commented on codersdk/deployment.go:
```diff
@@ -3420,6 +3423,17 @@ Write out the current server config as YAML to stdout.`,
+			Flag:        "workspace-agent-logs-retention",
+			Env:         "CODER_WORKSPACE_AGENT_LOGS_RETENTION",
+			Value:       &c.Retention.WorkspaceAgentLogs,
+			Default:     "7d",
```
What was your thought process for 7d? Is this table particularly egregious in its disk usage?

mafredri replied:
That was a copy paste comment from existing code, already fixed [`fe71946` (#21039)](https://github.com/coder/coder/pull/21039/commits/fe71946cf182e2477954e0ef6498fc9c12d3624e). 😉

I think 7d is fine considering the constraint:
- Latest build logs are _always_ kept
- Agent connection is considered

If you want to debug a build, it's quite likely that the logs will be present or old enough that you don't care.

Unfortunately, the original PR did not add much motivation for the value: https://github.com/coder/coder/pull/6558, but given this has been the value since it's inception, I'm OK with it staying this way.

mtojek reviewed (approved):
All good 👍

mafredri committed:
docs: clarify workspace agent logs retention is based on agent connection

The description was misleading by stating logs are deleted 'after this
period' without explaining that retention is based on when the agent
last connected, not the age of the logs themselves. Updated to clearly
state that logs from non-latest builds are deleted if the agent hasn't
connected within the retention period, and that logs from the latest
build are always retained.
